integration-test.Rmd

```diff
@@ -0,0 +1,330 @@
+
+```{r}
+# What value is currently assigned to the variable `portfolio_name_ref_all`?
+show_pattern_in_file <- function(file, pattern) {
+  grep(pattern, readLines(file), value = TRUE)
+}
+
+pattern <- "set_portfolio-name-ref-all_working-location_and_web-parameters.R"
+(files <- dir_ls("dry", regexp = pattern))
+
+this_pattern <- "portfolio_name_ref_all.*<-"
+matched <- map(files, show_pattern_in_file, pattern = this_pattern)
+walk(matched, writeLines)
+
+script_has_this_pattern <- grepl(this_pattern, matched)
+expect_true(any(script_has_this_pattern))
+```
+
+**NOTE: If the value of `portfolio_name_ref_all` comes from the user, we should provide an interface for the user to supply it -- so that the user needs not to change the source code.**
+
+## Configurations
+
+Ensure this configuration file exists:
+
+```{r}
+config_1 <- here(
+  "working_dir", 
+  "10_Parameter_File", 
+  "TestPortfolio_Input_PortfolioParameters.yml"
+)
+
+expect_true(file_exists(config_1))
+```
+
+```{r}
+look_into <- function(path, n = -1L) {
+  lines <- readLines(path, n, encoding = "UTF-8")
+  writeLines(lines)
+}
+
+look_into(config_1)
+```
+
+Ensure this other configuration file also exists:
+
+```{r}
+config_2 <- here("parameter_files", "WebParameters_2dii.yml")
+config_2_copy <- tempfile()
+fs::file_copy(config_2, config_2_copy)
+
+expect_true(file_exists(config_2))
+```
+
+Ensure the paths in the configuration file work both locally and on GitHub actions:
+
+```{r}
+make_config_portable <- function(config) {
+  lines <- readLines(config, encoding = "UTF-8")
+  lines <- make_paths_portable(lines)
+  writeLines(lines, config)
+  
+  invisible(config)
+}
+
+make_paths_portable <- function(x) {
+  x %>% 
+    root_field_path("project_location_ext", pattern = "PACTA_analysis") %>% 
+    root_field_path("data_location_ext", pattern = "pacta-data") %>% 
+    root_field_path("template_location", pattern = "create_interactive_report")
+}
+
+root_field_path <- function(x, field, pattern) {
+  parent <- path_dir(here())
+  value <- path(parent, extract_from(x, pattern))
+  sub(glue("({field}:[ ]?).*"), glue("\\1{value}/"), x)
+}
+
+extract_from <- function(x, pattern) {
+  line <- grep(pattern, x, value = TRUE)
+  sub(glue(".*({pattern}.*)"), "\\1", line)
+}
+
+make_config_portable(config_2)
+```
+
+```{r}
+config_paths <- config::get(file = config_2)$paths
+all_paths_exist <- all(map_lgl(config_paths, dir_exists))
+
+expect_true(all_paths_exist)
+```
+
+```{r}
+look_into(config_2)
+```
+
+## Scripts
+
+Populate the directory for processed inputs:
+
+```{r message=FALSE}
+dir_has_files <- function(path) {
+  stopifnot(is_dir(path))
+  
+  contents <- dir_ls(path, recurse = TRUE)
+  has_files <- any(map_lgl(contents, is_file))
+  has_files
+}
+
+out_1 <- path("working_dir", "30_Processed_Inputs")
+
+expect_false(dir_has_files(out_1))
+source("web_tool_script_1.R")
+expect_true(dir_has_files(out_1))
```

Comment:
I see this error (which surprisingly before used to be just a warning). Can we add the file "fund_data.fst" or remove the code that uses it?

```r
Loading PACTA.analysis
Quitting from lines 246-259 (integration-test.Rmd) 
Error: Error: '/home/runner/work/PACTA_analysis/PACTA_analysis/pacta-data/2019Q4/cleaned_files/fund_data.fst' does not exist.
Execution halted
```

Reply:
That depends on where the test reads that data from @maurolepore we can't add the file to the pacta-data gh repo, as it is more than 100mb in size. we could add it to the data folder locally though and the idea is that this data will be transferred to constructiva via sftp in the future, not via github. So in that sense, we can add it. But we might have to reflect these changes in the test and docker workflows..

Reply:
Thanks! And what should happen if the file isn't available? (Before I used to get a warning only -- not an error -- so I suspect there is a way for the analysis to continue without this file).

Reply:
there is indeed. the portfolio can also be processed without fund data. It is the same discussion we are having in the pacta2020 slack channel and @cjyetman opened a GH issue for this: https://github.com/2DegreesInvesting/PACTA_analysis/issues/123